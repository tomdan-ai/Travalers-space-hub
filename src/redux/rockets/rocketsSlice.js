import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  rockets: [],
  status: 'idle',
  error: null,
};

export const fetchRocketsData = createAsyncThunk('rockets/fetchRocketsData', async () => {
  try {
    const response = await axios.get('https://api.spacexdata.com/v4/rockets');
    const selectedData = response.data.map((rocket) => ({
      id: rocket.id,
      rocket_name: rocket.name,
      description: rocket.description,
      rocket_flickr_images: rocket.flickr_images,
    }));
    return selectedData;
  } catch (error) {
    throw new Error('Failed to fetch rocket data');
  }
});

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    reserveRocket: (state, action) => {
      const rocketId = action.payload;
      state.rockets = state.rockets.map((rocket) => (rocket.id === rocketId
        ? { ...rocket, reserved: true } : rocket));
    },
    unreserveRocket: (state, action) => {
      const rocketId = action.payload;
      state.rockets = state.rockets.map((rocket) => (rocket.id === rocketId
        ? { ...rocket, reserved: false } : rocket));
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchRocketsData.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchRocketsData.fulfilled, (state, { payload }) => {
        state.status = 'succeeded';
        state.rockets = payload;
      })
      .addCase(fetchRocketsData.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      });
  },

});

export const { reserveRocket, unreserveRocket } = rocketsSlice.actions;
export default rocketsSlice.reducer;
