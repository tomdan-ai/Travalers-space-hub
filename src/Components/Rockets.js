import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchRocketsData, reserveRocket, unreserveRocket } from '../redux/rockets/rocketsSlice';

const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.rockets);

  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(fetchRocketsData());
    }
  }, [rockets, dispatch]);

  const handleReserveRocket = (rocketId) => {
    const rocket = rockets.find((rock) => rock.id === rocketId);
    if (rocket.reserved) {
      dispatch(unreserveRocket(rocketId));
    } else {
      dispatch(reserveRocket(rocketId));
    }
  };

  return (
    <div>
      <ul className="rocket">
        {rockets.map((rocket) => (
          <li key={rocket.id}>
            <img src={rocket.rocket_flickr_images} alt={rocket.name} />
            <div>
              <h1>{rocket.rocket_name}</h1>
              <p>
                {rocket.reserved && <span>Reserved</span>}
                {rocket.description}
              </p>
              <button
                type="button"
                onClick={() => handleReserveRocket(rocket.id)}
                style={{
                  backgroundColor: rocket.reserved ? '#fff' : '#17bcee',
                  color: rocket.reserved ? '#000' : '#fff',
                  border: rocket.reserved ? '1px solid #000' : 'none',
                }}
              >
                {rocket.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
              </button>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Rockets;
