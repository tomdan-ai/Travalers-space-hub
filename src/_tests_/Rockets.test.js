import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Rockets from '../Components/Rockets';
import { fetchRocketsData, reserveRocket, unreserveRocket } from '../redux/rockets/rocketsSlice';

const mockStore = configureStore([]);

describe('Rockets component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      rockets: {
        rockets: [
          {
            id: 1,
            rocket_name: 'Falcon 1',
            description: 'A small rocket for small payloads.',
            rocket_flickr_images: 'https://www.flickr.com/photos/spacex/8884203235/in/photostream/',
            reserved: false,
          },
          {
            id: 2,
            rocket_name: 'Falcon 9',
            description: 'A medium-sized rocket for medium-sized payloads.',
            rocket_flickr_images: 'https://www.flickr.com/photos/spacex/8884203235/in/photostream/',
            reserved: true,
          },
        ],
      },
    });

    store.dispatch = jest.fn();
    store.dispatch(fetchRocketsData());

    component = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
  });

  it('dispatches fetchRocketsData action on mount', () => {
    expect(store.dispatch).toHaveBeenCalledWith(expect.any(Function));
  });

  it('dispatches reserveRocket action when Reserve Rocket button is clicked', () => {
    const reserveButton = component.getByText('Reserve Rocket');
    fireEvent.click(reserveButton);

    expect(store.dispatch).toHaveBeenCalledWith(reserveRocket(1));
  });

  it('dispatches unreserveRocket action when Cancel Reservation button is clicked', () => {
    const cancelButton = component.getByText('Cancel Reservation');
    fireEvent.click(cancelButton);

    expect(store.dispatch).toHaveBeenCalledWith(unreserveRocket(2));
  });

  it('matches the snapshot', () => {
    expect(component.container).toMatchSnapshot();
  });
});
