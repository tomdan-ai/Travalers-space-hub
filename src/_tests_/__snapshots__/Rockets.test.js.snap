// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets component matches the snapshot 1`] = `
<div>
  <div>
    <ul
      class="rocket"
    >
      <li>
        <img
          src="https://www.flickr.com/photos/spacex/8884203235/in/photostream/"
        />
        <div>
          <h1>
            Falcon 1
          </h1>
          <p>
            A small rocket for small payloads.
          </p>
          <button
            style="background-color: rgb(23, 188, 238); color: rgb(255, 255, 255);"
            type="button"
          >
            Reserve Rocket
          </button>
        </div>
      </li>
      <li>
        <img
          src="https://www.flickr.com/photos/spacex/8884203235/in/photostream/"
        />
        <div>
          <h1>
            Falcon 9
          </h1>
          <p>
            <span>
              Reserved
            </span>
            A medium-sized rocket for medium-sized payloads.
          </p>
          <button
            style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0); border: 1px solid #000;"
            type="button"
          >
            Cancel Reservation
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
